content/06_Transformation.md

```diff
@@ -0,0 +1,351 @@
+
+The second place I used `SelectMany` was slightly different: it was the final example of the [**Representing Filesystem Events in Rx** section in Chapter 3](./03_CreatingObservableSequences.md#representing-filesystem-events-in-rx). Although that example also combined multiple observable sources into a single observable, that list of observables was fixed: there was one for each of the different events from `FileSystemWatcher`. It used a different operator `Merge` (which we'll get to in [Combining Sequences](09_CombiningSequences.md)), which was simpler to use in that scenario because you just pass it the list of all the observables you'd like to combine. However, because of a few other things this code wanted to do (including deferred startup, automated disposal, and sharing a single source when multiple subscribers were active), the particular combination of operators used to achieve this meant our merging code that returned an `IObservable<FileSystemEventArgs>`, needed to be invoked as a transforming step. If we'd just used `Select`, the result would have been an `IObservable<IObservable<FileSystemEventArgs>>`. The structure of the code meant that it would only ever produce a single `IObservable<FileSystemEventArgs>`, so the double-wrapped type would be rather inconvenient. `SelectMany` is very useful in these scenarios—if composition of operators has introduced an extra layer of observables-in-observables that you don't want, `SelectMany` can unwrap one layer for you.
+
+These two cases—fanning out then back in, and removing or avoiding a layer of observables of observables—come up quite often, which makes `SelectMany` an important method. (It's not surprising that I was unable to avoid using it in earlier examples.)
+
+As it happens, `SelectMany` is also a particularly important operator in the mathematical theory that Rx is based on. It is a fundamental operator, in the sense that it is possible to build many other Rx operators with it. [Section 'Recreating other operators with `SelectMany`' in Appendix D](./D_AlgebraicUnderpinnings.md#recreating-other-operators-with-selectmany) shows how you can implement `Select` and `Where` using `SelectMany`.
+
+## Cast
+
+C#'s type system is not omniscient. Sometimes we might know something about the type of the values emerging from an observable source that is not reflected in that source's type. This might be based on domain-specific knowledge. For example, with the AIS message broadcast by ships, we might know that if the message type is 3, it will contain navigation information. That means we could write this:
```

Comment:
Does AIS use the parlance of ships or vessels? I say because there are often static platforms / sea planes etc

Reply:
This was a deliberate bit of weasel wording. I was avoiding the word "vessel" on purpose.

The fact that AIS messages can also be broadcast by things that are not ships is true, but doesn't make what I've written untrue. This paragraph is under no obligation to provide a complete description of everything that an AIS message conceivably could be. It's OK just to talk about what happens when a ship broadcasts an AIS message.

Given that it's OK to say "ship" why do I prefer that to "vessel" here? The short answer is: if you're looking for a concrete example (and I believe you are in favour of concrete examples) more concrete is generally better (for pretty much the same reason a concrete example is better than no concrete example).

It is true that AIS uses the more general language of "vessels". And if you're in a maritime frame of mind, that's clear enough. But I felt that for anyone who had landed directly at this part of the docs by searching, and wasn't already familiar with the AIS scenarios I use throughout, it might not be obvious what a "vessel" is.

It's _too_ general. Coffee cups are a kind of vessel.

I felt that "AIS messages broadcast by ships" would be more self-evidently clear to a wide audience.